tests/MassTransit.Tests/SagaStateMachineTests/CompositeEvent_Specs.cs

```diff
@@ -82,12 +82,12 @@ public TestStateMachine()
                     When(First)
                         .TransitionTo(WaitingForSecond));
 
-                CompositeEvent(() => Third, x => x.CompositeStatus, First, Second);
-
                 During(WaitingForSecond,
                     When(Third)
```

Comment:
The issue here is that the event is now being used before it's declared...

Reply:
True. However, the issue is that when we are looking for the states, **WaitingForSecond** has not yet been registered and can thus not be found when **CompositeEvent** is called.
![image](https://user-images.githubusercontent.com/15146247/151873425-e2c3284e-60f0-449a-8e00-104fe917f9a8.png)

Reply:
The _stateCache does contain the **WaitingForSecond** state, but it has no events associated yet.

Reply:
It should be, all the states are initialized at startup in the constructor.

Reply:
Yes, that is correct. Both the __stateCache_ and __eventCache_ are initialized during startup. However, the states in the __stateCache_ have no events associated yet. That is done when 'walking' the statemachine initialization.

Reply:
@phatboyg, I think I have found a solution. I am going to investigate a bit more tomorrow. It 10:55pm, time to sign off for today.

Reply:
Small update, I have all but one unittest working with the new setup. In the new setup CompositeEvents will have to be defined **before** Any _During_ statement. Instead of calling _During_ in the _CompositeEvent_ method I'm storing the bindings in a temporary dictionary which gets applied during _BindActivitiesToState_. So far, so good...